include/ygm/io/multi_output.hpp

```diff
@@ -0,0 +1,121 @@
+
+// Copyright 2019-2021 Lawrence Livermore National Security, LLC and other YGM
+// Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <fstream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <ygm/container/detail/hash_partitioner.hpp>
+
+namespace ygm::io {
+
+namespace fs = std::filesystem;
+
+template <typename Partitioner =
+              ygm::container::detail::hash_partitioner<std::string>>
+class multi_output {
+ public:
+  using self_type = multi_output<Partitioner>;
+
+  multi_output(ygm::comm &comm, const std::string &filename_prefix,
+               bool append = false)
+      : m_comm(comm),
+        pthis(this),
+        m_prefix_path(filename_prefix),
+        m_append_flag(append) {
+    pthis.check(m_comm);
+
+    fs::path prefix_path(filename_prefix);
+
+    // Create directories to hold files
+    if (comm.rank0()) {
+      make_directories(m_prefix_path);
+    }
+  }
+
+  ~multi_output() {
+    m_comm.barrier();
+    for (const auto &[fname, file_ptr] : m_map_file_pointers) {
+      file_ptr->close();
+    }
+  }
+
+  template <typename... Args>
+  void async_write_line(const std::string &filename, Args &&... args) {
```

Comment:
Shouldn't assume its an actual filename.   It could be a subpath to be appended to the prefixpath.   New directories might need to be created if subdirectory doesn't already exist.

Reply:
This is just a bad variable name. If the file corresponding to the subpath is not already open, `make_ofstream_ptr` gets called, which creates all missing subdirectories.